packages/react-router-dom/modules/Focus.js

```diff
@@ -0,0 +1,59 @@
+import React from "react";
+import { __RouterContext as RouterContext } from "react-router";
+import warning from "tiny-warning";
+
+class FocusWithLocation extends React.Component {
+  setRef = element => {
+    this.eleToFocus = element;
+  };
+
+  render() {
+    const { children } = this.props;
+    return children(this.setRef);
+  }
+
+  componentDidMount() {
+    this.focus();
+  }
+
+  componentDidUpdate(prevProps) {
+    // only re-focus when the location changes
+    if (this.props.location !== prevProps.location) {
```

Comment:
I can't find any great documentation for how focus should be handled when clicking a hash link. This could potentially take a function prop that lets the user do a more refined comparison.

````jsx
<Focus
  compare={(next, prev) => {
    // only re-focus when the pathname changes
    return next.pathname !== prev.pathname;
  }}
>...</Focus>
```

The best behavior would probably be to always re-focus and attach the `ref` to the component with the matching `id`. That requires a much more complex setup (conditionally attaching the `ref` to components for location's with a `hash` with a fallback to the "main" content when there is no `hash`), so I'm not sure how many people would actually go that far.

Reply:
This is the tricky part of this implementation IMO. I've never really considered object equality a good way to compare two `location` objects, so I always use `locationsAreEqual`. Having said that, this is a common thing that people do, so we should probably make the `location` object immutable between renders if it hasn't changed. Also, the `locationsAreEqual` function compares `location.state`, which I don't believe should trigger a refocus.

WRT the `compare` function, it feels like we should be able to pick the right thing to do based on what browsers currently do so we won't need to provide that hook. But I'm not 100% sure.

Reply:
I think that a reference check is okay here because it is the object emitted by `history`, but I'd be fine using `locationsAreEqual` instead.

I did a quick check of a static page and if you focus an element, then click a link, the focus is moved back to `document.body`. Maybe to start, `<Focus>` should re-focus for every location change. Down the road if people need more fine grained control, it would be easy enough to add the functionality.